crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -328,6 +328,24 @@ pub unsafe trait WorldQuery {
     /// constructing [`Self::Fetch`](crate::query::WorldQuery::Fetch).
     type State: Send + Sync + Sized;
 
+    /// This associated type defines how a query can be configured at runtime.
+    ///
+    /// [`QueryState::new_with_config`](crate::query::state::QueryState::new_with_config) uses this
+    /// associated type to construct [`QueryState`](crate::query::state::QueryState).
+    /// When this associated type implements the `Default` trait,
+    /// [`QueryState::new`](crate::query::state::QueryState::new) can be used instead.
+    ///
+    /// In the case where no config is required, `()` is conventionally chosen
+    /// as the type for `Config`. This is the case for statically defined
+    /// queries, for types such as `&T`, since they can obtain a
+    /// [`ComponentId`] without any runtime information.
+    ///
+    /// However, for dynamic queries, the [`ComponentId`] varies at runtime and
+    /// mut be additionally provided. By supplying `Config = ComponentId` we
+    /// can vary the component requested each time we construct a new query,
+    /// even for the same underlying Rust type.
```

Comment:
I'm trying to wrap my head around how this works but having a difficult time. Would it be worth adding a test or something so people like me can visualize it in action? 😅

A test would also allow us to verify this is working as intended for these dynamic usages.

Reply:
I don't have a test in this PR as it would require implementing a dynamic query from scratch which would be quite large.

@jakobhellermann has multiple example implementations, `Ptr, MutPtr, Vec`, in their PR:
https://github.com/jakobhellermann/bevy/blob/5617e3bb8337ae79e25445d468d274ec962d0305/crates/bevy_ecs/src/query/fetch.rs#L1546

I purposefully didn't include these implementations in this PR as it's not clear how to implement them in the "best" way, however, this PR allows users to at least implement them themselves :)